source/mesh_deformation/fastscapecc.cc

```diff
@@ -0,0 +1,698 @@
+        array_size = nx*ny;
+        
+        
+        
+        
+        
+        
+        
+        
+//
+//      dx = 20000;
+//      dy = dx ;
+//      x_extent = 100e3;
+//      y_extent = 100e3;
+//      nx = 5 ;
+//      ny = 5 ;
+//      array_size = nx*ny;
+//
+//      // Sub intervals are 3 less than points, if including the ghost nodes. Otherwise 1 less.
+//      table_intervals[0] = nx ;
+//      table_intervals[dim-1] = 1;
+//      table_intervals[1] = ny;
+
+//      const GeometryModel::Box<dim> *geometry
+//        = dynamic_cast<const GeometryModel::Box<dim>*> (&this->get_geometry_model());
+//
+//      // The first entry represents the minimum coordinates of the model domain, the second the model extent.
+//      for (unsigned int i=0; i<dim; ++i)
+//        {
+//          grid_extent[i].first = geometry->get_origin()[i];
+//          grid_extent[i].second = geometry->get_extents()[i];
+//        }
+    }
+
+
+    template <int dim>
+    void
+    FastScapecc<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                               AffineConstraints<double> &mesh_velocity_constraints,
+                                                               const std::set<types::boundary_id> &boundary_ids) const
+    {
+      if (this->get_timestep_number() == 0)
+        return;
+
+      // std::vector<std::vector<double>> temporary_variables;
+
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "FastScape plugin");
+
+      const unsigned int current_timestep = this->get_timestep_number ();
+
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      std::vector<std::vector<double>> temporary_variables(dim+2, std::vector<double>());
+
+      // Get a quadrature rule that exists only on the corners, and increase the refinement if specified.
+      const QIterated<dim-1> face_corners (QTrapez<1>(),
+                                           static_cast<unsigned int>(std::pow(2,additional_refinement+surface_refinement_difference)));
+
+      FEFaceValues<dim> fe_face_values (this->get_mapping(),
+                                        this->get_fe(),
+                                        face_corners,
+                                        update_values |
+                                        update_quadrature_points);
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned() && cell->at_boundary())
+          for (unsigned int face_no = 0; face_no < GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (cell->face(face_no)->at_boundary())
+              {
+                if ( cell->face(face_no)->boundary_id() != relevant_boundary)
+                  continue;
+
+                std::vector<Tensor<1,dim>> vel(face_corners.size());
```

Comment:
This is perhaps a more general comment:

The last line here is not very efficient. You are allocating memory on every face you encounter, but the array you allocate always has the same size. Two lines below you then fill it with some value. At the end of the code block, the memory is then released again.

A better scheme would have declared the variable just before the outer loop, because then the memory is allocated only once. Within the loop, you are then just filling the already allocated array, but the expensive memory allocation and deallocation only happens once in that case.